### PR TITLE
Azdevify dotnet podcasts

### PR DESCRIPTION
**Pull request by v-hongli1**
This pr is all changes to convert podcasts application into an Azd template.

You can currently deploy the podcasts application using azd by following the steps below:

1. azd up -t https://github.com/hongli750210/dotnet-podcasts.git -b issue#1055
2. Wait a moment for the resource deployment to complete, click the web endpoint and you will see:
image

@jongio, @rajeshkamal5050 for notification.

### CODE REVIEW

**Comment by jongio** [comment]
@hongli750210  - FYI you can also init with:


`azd init -t https://github.com/hongli750210/dotnet-podcasts.git -b issue#1055`

File: infra/app/hub.bicep [hunk 1]
```diff
@@ -0,0 +1,39 @@
+param name string
+param location string = resourceGroup().location
+param tags object = {}
+param appCommandLine string = ''
+param applicationInsightsName string = ''
+param appServicePlanId string
+param sqlServiceString string
+param orLeansStorage string
+@secure()
```
Comment (jongio) [inline infra/app/hub.bicep:9]: does this need to be secure?

File: infra/app/hub.bicep [hunk 2]
```diff
@@ -0,0 +1,39 @@
+param name string
+param location string = resourceGroup().location
+param tags object = {}
+param appCommandLine string = ''
+param applicationInsightsName string = ''
+param appServicePlanId string
+param sqlServiceString string
+param orLeansStorage string
```
Comment (jongio) [inline infra/app/hub.bicep:8]:
```suggestion
param orleansStorage string
```

Comment (jongio) [reply infra/app/hub.bicep:8]: You'll need to update the rest of the references to this var as well.

File: infra/app/hub.bicep [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    tags: union(tags, { 'azd-service-name': serviceName })
+    appCommandLine: appCommandLine
+    applicationInsightsName: applicationInsightsName
+    appServicePlanId: appServicePlanId
+    keyVaultName: keyVaultName
+    appSettings: {
+      REACT_APP_API_BASE_URL: podcastApiUrl
+      ASPNETCORE_ENVIRONMENT: 'Development'
+      REACT_LISTEN_TOGETHER_DB: sqlServiceString
+      REACT_OR_LEANS_STORAGE: orLeansStorage
```
Comment (jongio) [inline infra/app/hub.bicep:28]:
```suggestion
      REACT_ORLEANS_STORAGE: orLeansStorage
```

Comment (jongio) [reply infra/app/hub.bicep:28]: You'll need to update the other refs to this env var as well.

File: infra/app/api.bicep [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+param tags object = {}
+param applicationInsightsName string
+param appCommandLine string = ''
+param appServicePlanId string
+param sqlServiceString string
+param feedIngestion bool = false
+param allowedOrigins array = []
+param feedQueue string
+param keyVaultName string
+var serviceName = 'podcast.api'
```
Comment (jongio) [inline infra/app/api.bicep:12]: Add a blank line after the var definition.

File: infra/app/db.bicep [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+module sql '../core/database/sqlserver/sqlserver.bicep' = {
+  name: '${name}-sql-module'
+  params: {
+    name: name
+    location: location
+    databaseName: databaseName
+    keyVaultName: keyValutName
+    connectionStringKey: connectionStringKey
+    sqlAdminPassword: 'Pass@word'
+    appUserPassword: 'Pass@word'
```
Comment (jongio) [inline infra/app/db.bicep:18]: Please see the csharp-sql template for how we deal with passwords in templates and use that model.

File: infra/app/hub.bicep [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+param appCommandLine string = ''
+param applicationInsightsName string = ''
+param appServicePlanId string
+param sqlServiceString string
+param orLeansStorage string
+@secure()
+param podcastApiUrl string
+param keyVaultName string
+
+var serviceName = 'listentogether.hub'
```
Comment (jongio) [inline infra/app/hub.bicep:13]: Add blank line after var def.

File: infra/app/ingestion.bicep [hunk 7]
```diff
@@ -0,0 +1,33 @@
+param name string
+param location string = resourceGroup().location
+param tags object = {}
+param appCommandLine string = ''
+param applicationInsightsName string
+param appServicePlanId string
+param sqlServiceString string
+param feedQueue string
+param keyVaultName string
+var serviceName = 'podcast.ingestion.worker'
```
Comment (jongio) [inline infra/app/ingestion.bicep:10]: Blank line.  Please check other files for this as well and update them.

File: infra/app/storage.bicep [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    tags: union(tags, { 'azd-service-name': serviceName })
+    allowBlobPublicAccess: allowBlobPublicAccess
+  }
+}
+
+output name string = storage.outputs.name
+output id string = storage.outputs.id
+#disable-next-line outputs-should-not-contain-secrets
+output primaryKey string = storage.outputs.primaryKey
+output primaryEndpoints object = storage.outputs.primaryEndpoints
```
Comment (jongio) [inline infra/app/storage.bicep:21]: Let's figure out a way to mod the architecture to now have to output keys and connection strings.  I'm thinking we introduce a key vault.

File: infra/core/host/container-registry.bicep [hunk 9]
```diff
@@ -0,0 +1,36 @@
+param name string
```
Comment (jongio) [inline infra/core/host/container-registry.bicep:1]: Please only include files from /bicep/core that are used by the application.

File: infra/main.bicep [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+  scope: rg
+  params: {
+    name: '${abbrs.keyVaultVaults}${resourceToken}'
+    location: location
+    tags: tags
+    principalId: principalId
+  }
+}
+
+module apisql 'app/db.bicep' = {
```
Comment (jongio) [inline infra/main.bicep:86]:
```suggestion
module apiSql 'app/db.bicep' = {
```

File: infra/main.bicep [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+output REACT_APP_APPLICATIONINSIGHTS_CONNECTION_STRING string = monitoring.outputs.applicationInsightsConnectionString
+output REACT_APP_WEB_BASE_URL string = web.outputs.SERVICE_WEB_URI
+output REACT_APP_HUB_BASE_URL string = hub.outputs.SERVICE_WEB_URI
+output REACT_PODCASTS_DB string = apisql.outputs.DB_CONNECTION_STRING
+output REACT_LISTEN_TOGETHER_DB string = hubsql.outputs.DB_CONNECTION_STRING
+output REACT_FEED_QUEUE string = storage.outputs.FEED_QUEUE
+output REACT_OR_LEANS_STORAGE string = storage.outputs.OR_LEANS_STORAGE
+output FEATURES_FEED_INGESTION bool = feedIngestion
+output STORAGE_IMAGES string = storage.outputs.STORAGE_IMAGES
+output LISTEN_TOGETHER_HUB string = hub.outputs.LISTEN_TOGETHER_HUB
```
Comment (jongio) [inline infra/main.bicep:263]: Sort this alpha ascending - that does for all bicep files, thx.

**Comment by jongio** [comment]
The app doesn't load for me after running `azd up`

I'm seeing this in ingestion logs:

2023-01-03 20:39:49.381 +00:00 [Error] Middleware: Failed to forward request to http://169.254.129.4:8080. Encountered a System.Net.Http.HttpRequestException exception after 222.079ms with message: Connection refused. Check application logs to verify the application is properly handling HTTP traffic.

**Comment by v-xuto** [comment]
@jongio For related errors about running `azd up`, we are currently investigating and modifying the code. If there is any progress, We will update this PR.

**Comment by hongli750210** [comment]
@jongio 
For related errors about running `azd up`, after changing the deployed service from **App Service** to **Container App** in this PR, we no longer reproduce this issue when executing the command `azd up`.

However, we currently have a issue and need your help. 
After deployed in Azure, the displayed data on the subscribe discover page is not same as expected, and there is no event response when click the `Subscribe` button. As shown below:
image

- **Expected Result (deployed in local docker):**

image

After investigation, we found that during the `subscribe discover page` initialization, request forward to `<web-service-hostname>/show/{id}` and refresh the whole page. As shown below:
image

But the expected result is to forward to `<api-service-hostname>/shows/{id}` and refresh only part of the page. As shown below:
image

On Azure, the `Container App` for `Web`, `Api` and other services has been provided, and each service has its own host name. Therefore, when the page sends a request, the host names of the sender and receiver are inconsistent, resulting in misinterpretation of the route, resulting in The page is abnormal.
On the local docker, the docker containers configured for `Web`, `Api` and other services form a container group, and the `hostname` and `ports` are provided by this container group. So the request sent by the page will not be misinterpretation by the route.

We tried to provide a container group by bicep on Azure, but the following exception occurred:
image

After analyzing the exception information, we think that the prerequisite for the successful creation of the container group is that images are registered in the `Container Registry`, and it does not conform to the deployment logic of `azd up`.

Now we are blocked by this issue, do you have any ideas?

File: src/Services/ListenTogether/ListenTogether.Infrastructure/AzureDeveloperCliCredential.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
+﻿using Azure.Core;
```
Comment (jongio) [inline src/Services/ListenTogether/ListenTogether.Infrastructure/AzureDeveloperCliCredential.cs:1]: Please use AzureDeveloperCliCredential from the beta version of the Azure.Identity lib.

File: src/Services/Podcasts/Podcast.Infrastructure/AzureDeveloperCliCredential.cs [hunk 13]
```diff
@@ -0,0 +1,148 @@
+﻿using Azure.Core;
```
Comment (jongio) [inline src/Services/Podcasts/Podcast.Infrastructure/AzureDeveloperCliCredential.cs:1]: Also move to the new class in AzureIdentity preview

**Comment by jongio** [comment]
For the issue mentioned above it looks like the environment variables aren't being set in the app in Azure.  Did you double check that they are being set in the configuration section?

File: src/Services/Podcasts/Podcast.Infrastructure/AzureDeveloperCliCredential.cs (hunk 13 above)
Comment (hongli750210) [reply src/Services/Podcasts/Podcast.Infrastructure/AzureDeveloperCliCredential.cs]: As per your suggestion, fix the version of "Azure.Identity" to "1.9.0-beta.2" and use `AzureDeveloperCliCredential` from Azure.Identity library.

File: src/Services/ListenTogether/ListenTogether.Infrastructure/AzureDeveloperCliCredential.cs (hunk 12 above)
Comment (hongli750210) [reply src/Services/ListenTogether/ListenTogether.Infrastructure/AzureDeveloperCliCredential.cs]: As per your suggestion, fix the version of "Azure.Identity" to "1.9.0-beta.2" and use AzureDeveloperCliCredential from Azure.Identity library.

**Comment by jongio** [comment]
@hongli750210 - Could you look into why the actions are failing?

**Comment by hongli750210** [comment]
@jongio After checking the Credentials according to the prompt of the exception information of Action fail, the action fail may be caused by not installing `az cli` I think.

**Comment by jamesmontemagno** [comment]
Could just be you don't have admin to access the credentials here...

**Comment by jongio** [comment]
@jamesmontemagno - Do you want to try to kick off the action to see if you have perms to run it?

**Comment by jamesmontemagno** [comment]
I really don't know why it fails... works if I do a PR... mystery... let me create a new branch, merge this and then open a PR off a branch that @v-hongli1  can work from

**Pull request merged**